UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs

```diff
@@ -506,35 +564,40 @@ public Tuple<bool, ReagentMix> HasReagentsAtTile(Vector3Int localPosInt)
 
 	private void DistributeExcessToNearbyCells(ReagentMix excess, Vector3Int origin)
 	{
-		const int MAX_ITERATIONS = 250;
+		const int MAX_ITERATIONS = 450;
 		var iterations = 0;
-		var cellsToProcess = new Queue<Vector3Int>();
-		cellsToProcess.Enqueue(origin);
+		var cellsToProcess = new Queue<MetaDataNode>();
+		cellsToProcess.Enqueue(matrix.GetMetaDataNode(origin));
 		//(Max): This behavior for some reason breaks when the server is running at low FPS or heavily stuttring. (less than 20)
 		//I have no clue what's the cause, or how to mitgate this.
 		//You can emulate this issue by enabling gizmos in the editor and watching the FPS drop, then testing this.
 		while (cellsToProcess.Count > 0 && iterations < MAX_ITERATIONS)
 		{
 			var currentCell = cellsToProcess.Dequeue();
-			var neighbors = currentCell.GetNeighbors();
-			foreach (var neighbor in neighbors)
+			foreach (var neighbor in currentCell.Neighbors)
 			{
 				if (excess.Total <= 0) return;
 
 				// Skip if the neighbor is not passable
-				if (matrix.IsWallAt(neighbor, true)) continue;
-				var neighborReagents = HasReagentsAtTile(neighbor);
+				if (neighbor.IsOccupied) continue;
+				var neighborReagents = HasReagentsAtTile(neighbor.LocalPosition);
 
 				// If the neighboring cell is empty, add the excess and break up the amount
 				if (neighborReagents.Item1 == false)
 				{
-					tileReagentMixes.Add(neighbor, excess.Split(excess.Total / 2));
-					CreateLiquidOverlay(neighbor, excess);
+					tileReagentMixes.Add(neighbor.LocalPosition, excess.Split(excess.Total / 2));
+					CreateLiquidOverlay(neighbor.LocalPosition, excess);
 #if UNITY_EDITOR
-					GameGizmomanager.AddNewLineStaticClient(null, currentCell.ToWorldInt(matrix), null, neighbor.ToWorldInt(matrix), Color.red);
+					GameGizmomanager.AddNewLineStaticClient(null,
```

Comment:
maybe make this a setting on the  Meta layerdata like a bool on the inspector

Reply:
it's under a pre-processor, it's fine if its left like this for the time being.